hoot-rnd/src/main/cpp/hoot/rnd/cmd/MultiaryIngestCmd.cpp

```diff
@@ -0,0 +1,341 @@
+      _sortInput = false;
+      _writeChanges(_getNewInputStream(newInput), referenceOutput, changesetOutput);
+    }
+    else
+    {
+      LOG_INFO("Deriving a changeset between the input and reference data.");
+
+      //sort incoming data by element id, if necessary, for changeset derivation (only passing nodes
+      //through, so don't need to also sort by element type)
+      _sortedNewInput = _getSortedNewInput(newInput);
+
+      //create the changes and write them to the ref db layer and also to a changeset file for
+      //external use by Spark
+      _deriveAndWriteChanges(_getNewInputStream(_sortedNewInput), referenceOutput, changesetOutput);
+    }
+
+    LOG_INFO(
+      "Multiary data ingested from " << newInput << " into reference output layer: " <<
+      referenceOutput << " and changes written to changeset file: " << changesetOutput <<
+      ".  Time: " << FileUtils::secondsToDhms(_timer.elapsed()));
+    LOG_INFO("Changes Parsed: " << FileUtils::formatPotentiallyLargeNumber(_changesParsed));
+
+    return 0;
+  }
+
+private:
+
+  bool _sortInput;
+  QString _sortedNewInput;
+  bool _isGeoNamesInput;
+
+  long _changesParsed;
+
+  QElapsedTimer _timer;
+
+  QString _getSortedNewInput(const QString newInput)
+  {
+    if (!_sortInput)
+    {
+      return newInput;
+    }
+
+    LOG_DEBUG("Retrieving input stream...");
+
+    //write the unsorted input to a temp db layer; later it will be queried back out sorted by id
+
+    //TODO: if performance for this ends up being a bottleneck, will later implement something here
+    //that performs faster - either a file based merge sort, or using the db like this but with sql
+    //copy statements instead
+
+    boost::shared_ptr<PartialOsmMapReader> unsortedNewInputReader =
+      boost::dynamic_pointer_cast<PartialOsmMapReader>(
+        OsmMapReaderFactory::getInstance().createReader(newInput));
+    unsortedNewInputReader->setUseDataSourceIds(true);
+    unsortedNewInputReader->open(newInput);
+    boost::shared_ptr<ElementInputStream> unsortedNewInputStream =
+      boost::dynamic_pointer_cast<ElementInputStream>(unsortedNewInputReader);
+
+    LOG_DEBUG("Retrieving filtered input stream...");
+
+    boost::shared_ptr<ElementInputStream> filteredUnsortedNewInputStream =
+      _getFilteredNewInputStream(unsortedNewInputStream);
+
+    LOG_DEBUG("Retrieving output stream...");
+
+    const QString sortedNewInput =
+      HootApiDb::getBaseUrl().toString() + "/MultiaryIngest-tempNewInput-" +
+      QUuid::createUuid().toString();
+    boost::shared_ptr<HootApiDbWriter> unsortedNewInputWriter(new HootApiDbWriter());
+    unsortedNewInputWriter->setCreateUser(true);
+    unsortedNewInputWriter->setOverwriteMap(true);
+    unsortedNewInputWriter->open(sortedNewInput);
+    boost::shared_ptr<ElementOutputStream> unsortedNewOutputStream =
+      boost::dynamic_pointer_cast<ElementOutputStream>(unsortedNewInputWriter);
+
```

Comment:
I'm actually making things much harder than they need to be for sorting here by writing to a db temp layer.  Using the sort command on the geonames works just fine...was originally under the impression that it wasn't working...but it is.  Incidentally, it turns out that the geonames allCountries files are already sorted anyway (originally thought they weren't).  For PBF will simply check the sorted flag and fail out and if its unsorted.  Those are the only two formats that I'm supporting as input for this for now.  Any others can be added later.

Reply:
Yeah, its taking less than a minute to sort one of the randomized geonames country files vs. just over an hour to write it to the db for sorting.